### PR TITLE
Move total_estimated_cells from contents to top level (#2298)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Author (before primary review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes <sub>or this PR does not need porting</sub> 
- [x] Added `[r]` prefix at start of commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update` <sub>or this PR does not change requirements*.txt</sub>
- [x] Mentioned `make requirements` in UPGRADING.rst <sub>or this PR does not change requirements*.txt</sub>
- [x] Documented upgrade of personal deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `[r]` prefix of commit title
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox` <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev` <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

**Review by jessebrennan — APPROVED** [review]
This looks good to me :+1:

File: src/azul/indexer/document.py [hunk 1]
```diff
@@ -553,6 +553,7 @@ def to_source(self):
 class Aggregate(Document[AggregateCoordinates]):
     bundles: Optional[List[JSON]]
     num_contributions: int
+    total_estimated_cells: int
```
Comment (hannes-ucsc) [inline src/azul/indexer/document.py:556]: This is an HCA specific that must be isolated in the plugin.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
We worked so hard on the plug-in abstraction, we're not going to throw that away by reintroducing HCA specifics into the indexer.

File: src/azul/indexer/document_service.py [hunk 2]
```diff
@@ -77,11 +76,12 @@ def field_types(self, catalog: CatalogName) -> FieldTypes:
         :return: dict with nested keys matching Elasticsearch fields and values with the field's type
         """
         field_types = {}
+        aggregate_class = self.metadata_plugin(catalog).aggregate_class()
```
Comment (hannes-ucsc) [inline src/azul/indexer/document_service.py:79]:
```suggestion
        aggregate_cls = self.metadata_plugin(catalog).aggregate_class()
```

File: src/azul/indexer/index_service.py [hunk 3]
```diff
@@ -420,17 +420,17 @@ def _aggregate(self, contributions: List[CataloguedContribution]) -> List[Aggreg
...
                 dict(uuid=c.coordinates.bundle.uuid,
                      version=c.coordinates.bundle.version)
                 for c in contributions
             ]
-            aggregate = Aggregate(coordinates=AggregateCoordinates(entity=entity),
-                                  version=None,
-                                  contents=contents,
-                                  bundles=bundles,
-                                  num_contributions=tallies[entity])
+            aggregate_class = self.metadata_plugin(entity.catalog).aggregate_class()
```
Comment (hannes-ucsc) [inline src/azul/indexer/index_service.py:428]:
```suggestion
            aggregate_cls = self.metadata_plugin(entity.catalog).aggregate_class()
```

File: src/azul/indexer/transform.py [hunk 4]
```diff
@@ -71,11 +68,3 @@ def get_aggregator(cls, entity_type) -> EntityAggregator:
         entities of types other than X.
         """
         raise NotImplementedError
-
-    @classmethod
-    def post_process_aggregate(cls, contents: MutableJSON) -> None:
```
Comment (hannes-ucsc) [inline src/azul/indexer/transform.py:76]: Glad this is gone.

File: src/azul/plugins/__init__.py [hunk 5]
```diff
@@ -129,6 +132,13 @@ def service_config(self) -> ServiceConfig:
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def aggregate_class(self) -> Type['Aggregate']:
```
Comment (hannes-ucsc) [inline src/azul/plugins/__init__.py:136]:
```suggestion
    def aggregate_class(self) -> Type[Aggregate]:
```

https://www.python.org/dev/peps/pep-0484/#forward-references

File: src/azul/plugins/__init__.py [hunk 6]
```diff
@@ -129,6 +132,13 @@ def service_config(self) -> ServiceConfig:
...
+        """
+        Returns the MetadataPlugin specific Aggregate class
```
Comment (hannes-ucsc) [inline src/azul/plugins/__init__.py:138]:
```suggestion
        Returns the concrete class to use for representing aggregate documents in the indexer.
```

and wrap properly

File: src/azul/plugins/metadata/hca/aggregate.py [hunk 7]
```diff
@@ -24,11 +28,41 @@
...
 )
 
 
+@dataclass
+class HCAAggregate(Aggregate):
+    total_estimated_cells: int = field(init=False)
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.total_estimated_cells = self._total_estimated_cells()
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/aggregate.py:47]: Seems overly complicated. Can't you use our default mechanism for lazily computed values: @cached_property?

File: test/indexer/test_hca_indexer.py [hunk 8]
```diff
@@ -162,6 +171,7 @@ def test_deletion(self):
         self.assertTrue(min(bundle_sizes.values()) < IndexWriter.bulk_threshold < max(bundle_sizes.values()))
 
         field_types = self.index_service.catalogued_field_types()
+        aggregate_class = self.metadata_plugin().aggregate_class()
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:174]:
```suggestion
        aggregate_cls = self.metadata_plugin().aggregate_class()
```

File: test/indexer/test_hca_indexer.py [hunk 9]
```diff
@@ -138,6 +143,10 @@ def tearDown(self):
     translated_bool_true = null_bool.to_index(True)
     translated_bool_false = null_bool.to_index(False)
 
+    @lru_cache(maxsize=None)
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:146]: It looks like you are uncertain when to use @cached_property and when to use @lru_cache. Feel free to ask after stand-up.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/indexer/document.py [hunk 10]
```diff
@@ -436,8 +436,12 @@ def mandatory_source_fields(cls) -> List[str]:
         return ['entity_id']
 
     @classmethod
-    def from_index(cls, field_types: CataloguedFieldTypes, hit: JSON) -> 'Document':
-        coordinates = DocumentCoordinates.from_hit(hit)
+    def from_index(cls,
+                   field_types: CataloguedFieldTypes,
+                   hit: JSON,
+                   coordinates: Optional[DocumentCoordinates[CataloguedEntityReference]] = None) -> 'Document':
```
Comment (hannes-ucsc) [inline src/azul/indexer/document.py:442]:
```suggestion
                   *,
                   coordinates: Optional[DocumentCoordinates[CataloguedEntityReference]] = None) -> 'Document':
```

File: src/azul/indexer/index_service.py [hunk 11]
```diff
@@ -287,16 +287,24 @@ def _read_aggregates(self, entities: CataloguedTallies) -> Dict[CataloguedEntity
                 for coordinate in coordinates
             ]
         }
+        mandatory_source_fields = set()
+        for coordinate in coordinates:
+            aggregate_cls = self.metadata_plugin(coordinate.entity.catalog).aggregate_class()
```
Comment (hannes-ucsc) [inline src/azul/indexer/index_service.py:292]: I think we should first collect all catalogs in a set and then iterate over that set to populate mandatory_source_fields. 

We may also want to extract `self.metadata_plugin(catalog).aggregate_class()` into a method and think about applying lru_cache to that method.

File: src/azul/indexer/index_service.py [hunk 12]
```diff
@@ -287,16 +287,24 @@ def _read_aggregates(self, entities: CataloguedTallies) -> Dict[CataloguedEntity
...
+
+        def aggregates():
+            for doc in response['docs']:
+                if doc['found']:
+                    coordinate = DocumentCoordinates.from_hit(doc)
+                    aggregate_cls = self.metadata_plugin(coordinate.entity.catalog).aggregate_class()
+                    aggregate = aggregate_cls.from_index(self.catalogued_field_types(), doc, coordinate)
+                    yield aggregate
+
         return {
```
Comment (hannes-ucsc) [inline src/azul/indexer/index_service.py:305]: See if this fits into one line.

File: src/azul/plugins/metadata/hca/aggregate.py [hunk 13]
```diff
@@ -24,11 +30,37 @@
...
+
+    @cached_property
+    def total_estimated_cells(self) -> int:
+        return sum(cs['total_estimated_cells']
+                   for cs in self.contents['cell_suspensions']
+                   if cs['total_estimated_cells'] is not None)
+
+    @classmethod
+    def field_types(cls, field_types: FieldTypes) -> FieldTypes:
+        return {
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/aggregate.py:54]: Literal …

File: src/azul/plugins/metadata/hca/aggregate.py [hunk 14]
```diff
@@ -24,11 +30,37 @@
...
+            **super().field_types(field_types),
+            'total_estimated_cells': pass_thru_int
+        }
+
+    def to_source(self) -> JSON:
+        return dict(super().to_source(),
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/aggregate.py:60]: or `dict` constructor. Chose one. I think the latter.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by danielsotirhos** [comment]
Please advise if you have an idea how this PyCharm warning could be resolved.
`aggregate_class()` is type hinted to return a `Type[Aggregate]` however PyCharm doesn't seem to understand.

Screen Shot 2020-10-28 at 9 03 40 AM

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
For demo, show the field in the index, using Kibana, and demonstrate that existing functionality (sorting/filtering) is not impeded.

**Comment by hannes-ucsc** [comment]
The reindex on dev had some issues caused by PR #2381 for #2330: 600 or so failed tallies. Nevertheless, this can still be demoed and closed.

**Pull request merged**